xray/plot/plot.py

```diff
@@ -257,7 +260,55 @@ def _determine_cmap_params(plot_data, vmin=None, vmax=None, cmap=None,
         else:
             extend = 'neither'
 
-    return vmin, vmax, cmap, extend
+    if levels:
+        cmap, cnorm = _determine_discrete_cmap_params(cmap, levels,
+                                                      vmin, vmax,
+                                                      extend)
+    else:
+        cnorm = None
+
+    return vmin, vmax, cmap, extend, cnorm
+
+
+def _determine_discrete_cmap_params(cmap, levels, vmin, vmax, extend):
+    """
+    Build a descrete colormap and normalization of the data.
+    """
+    import matplotlib as mpl
+    import matplotlib.pyplot as plt
+
+    if extend == 'both':
+        ext_n = 2
+    elif extend in ['min', 'max']:
+        ext_n = 1
+    else:
+        ext_n = 0
+
+    if isinstance(levels, int):
+        cticks = np.linspace(vmin, vmax, num=levels + 1, endpoint=True)
+        n_colors = levels + ext_n
+    elif isinstance(levels, Sequence):
+        cticks = np.asarray(levels)
+        n_colors = len(levels) + ext_n - 1
+    else:
+        TypeError('Unexpected type (%s) given for levels' % type(levels))
+
+    try:
+        from seaborn.apionly import color_palette
+        pal = color_palette(cmap, n_colors=n_colors)
```

Comment:
This section here could use a bit more thought.  The nice part about using seaborn's `color_palette` is that it allows us to access all its color schemes. 

There are two downsides:

1.  We add seaborn as a dependency or hope to recover correctly (As shown below)
2.  `color_palette` doesn't work with `LinearSegmentedColormap` objects which is how `viridis` is currently stored.